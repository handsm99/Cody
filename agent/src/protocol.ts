/* eslint-disable @typescript-eslint/consistent-type-definitions */
import { RecipeID } from '@sourcegraph/cody-shared/src/chat/recipes/recipe'
import { ChatMessage } from '@sourcegraph/cody-shared/src/chat/transcript/messages'
import { event } from '@sourcegraph/cody-shared/src/sourcegraph-api/graphql/client'

import { CompletionEvent } from '../../vscode/src/completions/logger'

// This file documents the Cody Agent JSON-RPC protocol. Consult the JSON-RPC
// specification to learn about how JSON-RPC works https://www.jsonrpc.org/specification
// The Cody Agent server only supports transport via stdout/stdin.

// The JSON-RPC requests of the Cody Agent protocol. Requests are async
// functions that return some (possibly null) value.
export type Requests = {
    // ================
    // Client -> Server
    // ================

    // The 'initialize' request must be sent at the start of the connection
    // before any other request/notification is sent.
    initialize: [ClientInfo, ServerInfo]
    // The 'shutdown' request must be sent before terminating the agent process.
    shutdown: [null, null]

    // Client requests the agent server to lists all recipes that are supported
    // by the agent.
    'recipes/list': [null, RecipeInfo[]]
    // Client requests the agent server to execute an individual recipe.
    // The response is null because the AI/Assistant messages are streamed through
    // the chat/updateMessageInProgress notification. The flow to trigger a recipe
    // is like this:
    // client --- recipes/execute --> server
    // client <-- chat/updateMessageInProgress --- server
    //             ....
    // client <-- chat/updateMessageInProgress --- server
    'recipes/execute': [ExecuteRecipeParams, null]

    'autocomplete/execute': [AutocompleteParams, AutocompleteResult]

    'graphql/currentUserId': [null, string]
    'graphql/logEvent': [event, null]

    'graphql/getRepoIdIfEmbeddingExists': [{ repoName: string }, string | null]
    'graphql/getRepoId': [{ repoName: string }, string | null]

    // ================
    // Server -> Client
    // ================
}

// The JSON-RPC notifications of the Cody Agent protocol. Notifications are
// synchronous fire-and-forget messages that have no return value. Notifications are
// conventionally used to represent streams of values.
export type Notifications = {
    // ================
    // Client -> Server
    // ================

    // The 'initalized' notification must be sent after receiving the 'initialize' response.
    initialized: [null]
    // The 'exit' notification must be sent after the client receives the 'shutdown' response.
    exit: [null]

    // The server should use the provided connection configuration for all
    // subsequent requests/notifications. The previous extension configuration
    // should no longer be used.
    'extensionConfiguration/didChange': [ExtensionConfiguration]

    // Lifecycle notifications for the client to notify the server about text
    // contents of documents and to notify which document is currently focused.
    'textDocument/didOpen': [TextDocument]
    // The 'textDocument/didChange' notification should be sent on almost every
    // keystroke, whether the text contents changed or the cursor/selection
    // changed.  Leave the `content` property undefined when the document's
    // content is unchanged.
    'textDocument/didChange': [TextDocument]
    // The user focused on a document without changing the document's content.
    // Only the 'uri' property is required, other properties are ignored.
    'textDocument/didFocus': [TextDocument]
    // The user closed the editor tab for the given document.
    // Only the 'uri' property is required, other properties are ignored.
    'textDocument/didClose': [TextDocument]

    '$/cancelRequest': [CancelParams]
    // The user no longer wishes to consider the last autocomplete candidate
    // and the current autocomplete id should not be reused.
    'autocomplete/clearLastCandidate': [null]

    // Resets the chat transcript and clears any in-progress interactions.
    // This notification should be sent when the user starts a new conversation.
    // The chat transcript grows indefinitely if this notification is never sent.
    'transcript/reset': [null]

    // ================
    // Server -> Client
    // ================
    // The server received new messages for the ongoing 'chat/executeRecipe'
    // request. The server should never send this notification outside of a
    // 'chat/executeRecipe' request.
    'chat/updateMessageInProgress': [ChatMessage | null]

    'debug/message': [DebugMessage]
}

export interface CancelParams {
    id: string | number
}

export interface AutocompleteParams {
    filePath: string
    position: Position
    // Defaults to 'Automatic' for autocompletions which were not explicitly
    // triggered.
    triggerKind?: 'Automatic' | 'Invoke'
}

export interface AutocompleteResult {
    items: AutocompleteItem[]
    completionEvent?: CompletionEvent
}

export interface AutocompleteItem {
    insertText: string
    range: Range
}

export interface ClientInfo {
    name: string
    version: string
    workspaceRootUri: string

    /** @deprecated Use `workspaceRootUri` instead. */
    workspaceRootPath?: string

    /** @deprecated Use `extensionConfiguration` instead. */
    connectionConfiguration?: ExtensionConfiguration
    extensionConfiguration?: ExtensionConfiguration
    capabilities?: ClientCapabilities
}

export interface ClientCapabilities {
    completions?: 'none'
    //  When 'streaming', handles 'chat/updateMessageInProgress' streaming notifications.
    chat?: 'none' | 'streaming'
}

export interface ServerInfo {
    name: string
    authenticated: boolean
    codyEnabled: boolean
    codyVersion: string | null
    capabilities?: ServerCapabilities
}
export interface ServerCapabilities {}

export interface ExtensionConfiguration {
    serverEndpoint: string
    proxy?: string | null
    accessToken: string
    customHeaders: Record<string, string>
    autocompleteAdvancedProvider?: string
    autocompleteAdvancedServerEndpoint?: string | null
    autocompleteAdvancedModel?: string | null
    autocompleteAdvancedAccessToken?: string | null
    debug?: boolean
    verboseDebug?: boolean
    codebase?: string
}

export interface Position {
    // 0-indexed
    line: number
    // 0-indexed
    character: number
}

export interface Range {
    start: Position
    end: Position
}

export interface TextDocument {
    filePath: string
    content?: string
    selection?: Range
}

export interface RecipeInfo {
    id: RecipeID
    title: string // Title Case
}

export interface ExecuteRecipeParams {
    id: RecipeID
    humanChatInput: string
    data?: any
}

export interface DebugMessage {
    channel: string
    message: string
}
