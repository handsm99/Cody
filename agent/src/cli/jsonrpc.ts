import NodeHttpAdapter from '@pollyjs/adapter-node-http'
import { Polly, type EXPIRY_STRATEGY, type MODE, type Request } from '@pollyjs/core'
import { type Har } from '@pollyjs/persister'
import FSPersister from '@pollyjs/persister-fs'
import * as commander from 'commander'
import { Command, Option } from 'commander'

import { Agent } from '../agent'

import { decodeCompressedBase64 } from './base64'
import { booleanOption } from './evaluate-autocomplete/cli-parsers'
import { PollyYamlWriter } from './pollyapi'

interface JsonrpcCommandOptions {
    expiresIn?: string | null | undefined
    recordingDirectory?: string
    recordingMode?: MODE
    recordIfMissing?: boolean
    recordingExpiryStrategy?: EXPIRY_STRATEGY
    recordingName?: string
}

function recordingModeOption(value: string): MODE {
    switch (value) {
        case 'record':
        case 'replay':
        case 'passthrough':
        case 'stopped':
            return value
        default:
            throw new commander.InvalidArgumentError(
                'Not a valid recording mode. Valid options are record, replay-or-record, replay, passthrough, or stopped.'
            )
    }
}

function expiryStrategyOption(value: string): EXPIRY_STRATEGY {
    switch (value) {
        case 'error':
        case 'warn':
        case 'record':
            return value
        default:
            throw new commander.InvalidArgumentError(
                'Not a valid expiry strategy. Valid options are error, warn, or record.'
            )
    }
}

class CodyNodeHttpAdapter extends NodeHttpAdapter {
    public async onRequest(request: Request): Promise<void> {
        if (request.body) {
            request.body = request.body
                .replaceAll(/`([^`]*)(cody-vscode-shim-test[^`]*)`/g, '`$2`')
                .replaceAll(/(\\\\)(\w)/g, '/$2')
        }

        return super.onRequest(request)
    }
}

/**
 * The default file system persister with the following customizations
 *
 * - Replaces Cody access tokens with the string "REDACTED" because we don't
 *   want to commit the access token into git.
 * - To avoid diff churn/conflicts:
 *   - Sets date headers to a known static date
 *   - Removes cookies
 *   - Sets dates/timing information stored by Polly to static values
 */
class CodyPersister extends FSPersister {
    // HACK: `FSPersister` has a private `api` property that writes the
    // recording.har file using JSON format. We override the `api` property here
    // with a custom implementation that uses YAML format instead. This property
    // is intentionally marked as public even if it's not used anywhere.
    public api: PollyYamlWriter

    constructor(polly: any) {
        super(polly)
        if (!this.options.recordingsDir) {
            throw new Error('No recording directory provided')
        }
        this.api = new PollyYamlWriter(this.options.recordingsDir)
    }
    public static get id(): string {
        return 'cody-fs'
    }

    public async onFindRecording(recordingId: string): Promise<Har | null> {
        const har = await super.onFindRecording(recordingId)
        if (har === null) {
            return har
        }
        for (const entry of har.log.entries) {
            const postData = entry?.request?.postData
            if (postData !== undefined && postData?.text === undefined && (postData as any)?.textJSON !== undefined) {
                // Format `postData.textJSON` back into the escaped string for the `.text` format.
                postData.text = JSON.stringify((postData as any).textJSON)
                ;(postData as any).textJSON = undefined
            }
        }
        return har
    }

    public onSaveRecording(recordingId: string, recording: Har): Promise<void> {
        const entries = recording.log.entries
        recording.log.entries.sort((a, b) => a.request.url.localeCompare(b.request.url))
        for (const entry of entries) {
            if (entry.request?.postData?.text?.startsWith('{')) {
                // Format `postData.text` as a JSON object instead of escaped string.
                // This makes it much easier to review the har file locally.
                const postData: any = entry.request.postData
                postData.textJSON = JSON.parse(entry.request.postData.text)
                postData.text = undefined
            }
            // Clean up the entries to reduce the size of the diff when re-recording
            // and to remove any access tokens.

            // Update any headers
            entry.request.bodySize = undefined
            entry.request.headersSize = 0
            entry.response.bodySize = undefined
            entry.response.headersSize = 0
            const headers = [...entry.request.headers, ...entry.response.headers]
            for (const header of headers) {
                switch (header.name) {
                    case 'authorization':
                        header.value = 'token REDACTED'
                        break
                    case 'date':
                        header.value = 'Fri, 05 Jan 2024 11:11:11 GMT'
                        break
                    case 'retry-after':
                        header.value = '0'
                        break
                }
            }

            // Remove any headers and cookies we don't need at all.
            entry.request.headers = this.filterHeaders(entry.request.headers)
            entry.response.headers = this.filterHeaders(entry.response.headers)
            entry.request.cookies.length = 0
            entry.response.cookies.length = 0

            // And other misc fields.
            entry.startedDateTime = 'Fri, 05 Jan 2024 00:00:00 GMT'
            entry.time = 0
            entry.timings = {
                blocked: -1,
                connect: -1,
                dns: -1,
                receive: 0,
                send: 0,
                ssl: -1,
                wait: 0,
            }
            const responseContent = entry.response.content
            if (
                responseContent?.encoding === 'base64' &&
                responseContent?.mimeType === 'application/json' &&
                responseContent.text
            ) {
                // The GraphQL responses are base64+gzip encoded. We decode them
                // in a sibling `textDecoded` property so we can more easily review
                // in in pull requests.
                try {
                    const text = JSON.parse(responseContent.text)[0]
                    const decodedBase64 = decodeCompressedBase64(text)
                    ;(responseContent as any).textDecoded = decodedBase64
                } catch {
                    // Ignored: uncomment below to debug. It's fine to ignore this error because we only
                    // make a best-effort to decode the gzip+base64 encoded JSON payload. It's not needed
                    // for the HTTP replay to work correctly because we leave the `.text` property unchanged.
                    // console.error('base64 decode error', error)
                }
            }
        }
        return super.onSaveRecording(recordingId, recording)
    }

    private filterHeaders(headers: { name: string; value: string }[]): { name: string; value: string }[] {
        const removeHeaderNames = new Set(['set-cookie', 'server', 'via'])
        const removeHeaderPrefixes = ['x-trace', 'cf-']
        return headers.filter(
            header =>
                !removeHeaderNames.has(header.name) &&
                removeHeaderPrefixes.every(prefix => !header.name.startsWith(prefix))
        )
    }
}

export const jsonrpcCommand = new Command('jsonrpc')
    .description(
        'Interact with the Agent using JSON-RPC via stdout/stdin. ' +
            'This is the subcommand that is used by Cody clients like the JetBrains and Neovim plugins.'
    )
    .addOption(
        new Option(
            '--recording-directory <path>',
            'Path to the directory where network traffic is recorded or replayed from. This option should only be used in testing environments.'
        ).env('CODY_RECORDING_DIRECTORY')
    )
    .addOption(
        new Option(
            '--recording-mode <mode>',
            'What kind of recording mode to use. Valid values are to the directory where network traffic is recorded or replayed from. This option should only be used in testing environments.'
        )
            .argParser(recordingModeOption)
            .env('CODY_RECORDING_MODE')
    )
    .addOption(
        new Option(
            '--recording-name <mode>',
            'The name of the recording to use. Every unique name results in a unique recording (HAR file). Use a unique name for every unique test in your test suite.'
        ).env('CODY_RECORDING_NAME')
    )
    .addOption(
        new Option(
            '--recording-expiry-strategy <strategy>',
            'What to do when encountering an expired recording). Use a unique name for every unique test in your test suite.'
        )
            .argParser(expiryStrategyOption)
            .env('CODY_RECORDING_EXPIRY_STRATEGY')
            .default('error')
    )
    .addOption(
        new Option('--recording-expires-in <duration>', 'When to expire the recordings')
            .env('CODY_RECORDING_EXPIRES_IN')
            .default('365d')
    )
    .addOption(
        new Option('--record-if-missing <true|false>', 'If false, fails the test instead of recording')
            .env('CODY_RECORD_IF_MISSING')
            .argParser(booleanOption)
            .default(false)
    )
    .action((options: JsonrpcCommandOptions) => {
        let polly: Polly | undefined
        if (options.recordingDirectory) {
            if (options.recordingMode === undefined) {
                console.error('CODY_RECORDING_MODE is required when CODY_RECORDING_DIRECTORY is set.')
                process.exit(1)
            }
            Polly.register(CodyNodeHttpAdapter)
            Polly.register(CodyPersister)
            polly = new Polly(options.recordingName ?? 'CodyAgent', {
                flushRequestsOnStop: true,
                recordIfMissing: options.recordIfMissing ?? options.recordingMode === 'record',
                mode: options.recordingMode,
                adapters: ['node-http'],
                persister: 'cody-fs',
                recordFailedRequests: true,
                expiryStrategy: options.recordingExpiryStrategy,
                expiresIn: options.expiresIn,
                persisterOptions: {
                    keepUnusedRequests: true,
                    fs: {
                        recordingsDir: options.recordingDirectory,
                    },
                },
                matchRequestsBy: {
                    headers: false,
                    order: false,
                },
            })
            // Automatically pass through requests to download bfg binaries
            // because we don't want to record the binary downloads for
            // macOS/Windows/Linux
            polly.server.get('https://github.com/sourcegraph/bfg/*path').passthrough()
            polly.server.get('https://github.com/sourcegraph/symf/*path').passthrough()
        } else if (options.recordingMode) {
            console.error('CODY_RECORDING_DIRECTORY is required when CODY_RECORDING_MODE is set.')
            process.exit(1)
        }

        if (process.env.CODY_DEBUG === 'true') {
            process.stderr.write('Starting Cody Agent...\n')
        }

        const agent = new Agent({ polly })

        // Force the agent process to exit when stdin/stdout close as an attempt to
        // prevent zombie agent processes. We experienced this problem when we
        // forcefully exit the IntelliJ process during local `./gradlew :runIde`
        // workflows. We manually confirmed that this logic makes the agent exit even
        // when we forcefully quit IntelliJ
        // https://github.com/sourcegraph/cody/pull/1439#discussion_r1365610354
        process.stdout.on('close', () => process.exit(1))
        process.stdin.on('close', () => process.exit(1))

        process.stdin.pipe(agent.messageDecoder)
        agent.messageEncoder.pipe(process.stdout)
    })
