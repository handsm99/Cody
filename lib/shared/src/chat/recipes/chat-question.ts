import { CodebaseContext } from '../../codebase-context'
import { ContextMessage, getContextMessageWithResponse } from '../../codebase-context/messages'
import { Editor, SelectionText, TextDocument, uriToPath } from '../../editor'
import { DocumentOffsets } from '../../editor/offsets'
import { IntentDetector } from '../../intent-detector'
import { MAX_CURRENT_FILE_TOKENS, MAX_HUMAN_INPUT_TOKENS } from '../../prompt/constants'
import {
    populateCurrentEditorContextTemplate,
    populateCurrentEditorSelectedContextTemplate,
} from '../../prompt/templates'
import { truncateText } from '../../prompt/truncation'
import { Interaction } from '../transcript/interaction'

import { isSingleWord, numResults } from './helpers'
import { Recipe, RecipeContext, RecipeID } from './recipe'

export class ChatQuestion implements Recipe {
    public id: RecipeID = 'chat-question'

    constructor(private debug: (filterLabel: string, text: string, ...args: unknown[]) => void) {}

    public async getInteraction(humanChatInput: string, context: RecipeContext): Promise<Interaction | null> {
        const truncatedText = truncateText(humanChatInput, MAX_HUMAN_INPUT_TOKENS)
        return Promise.resolve(
            new Interaction(
                { speaker: 'human', text: truncatedText, displayText: humanChatInput },
                { speaker: 'assistant' },
                this.getContextMessages(
                    truncatedText,
                    context.editor,
                    context.firstInteraction,
                    context.intentDetector,
                    context.codebaseContext,
                    context.editor.getActiveTextDocumentSelectionText()
                ),
                []
            )
        )
    }

    private async getContextMessages(
        text: string,
        editor: Editor,
        firstInteraction: boolean,
        intentDetector: IntentDetector,
        codebaseContext: CodebaseContext,
        selection: SelectionText | null
    ): Promise<ContextMessage[]> {
        const contextMessages: ContextMessage[] = []
        // If input is less than 2 words, it means it's most likely a statement or a follow-up question that does not require additional context
        // e,g. "hey", "hi", "why", "explain" etc.
        const isTextTooShort = isSingleWord(text)
        if (isTextTooShort) {
            return contextMessages
        }

        const isCodebaseContextRequired = firstInteraction || (await intentDetector.isCodebaseContextRequired(text))

        this.debug('ChatQuestion:getContextMessages', 'isCodebaseContextRequired', isCodebaseContextRequired)
        if (isCodebaseContextRequired) {
            const codebaseContextMessages = await codebaseContext.getCombinedContextMessages(text, numResults)
            contextMessages.push(...codebaseContextMessages)
            // TODO: Put this in getContextMessages?
            // Let's leave this here for now :)
            contextMessages.push(...(await codebaseContext.getGraphContextMessages()))
        }

        const isEditorContextRequired = intentDetector.isEditorContextRequired(text)
        this.debug('ChatQuestion:getContextMessages', 'isEditorContextRequired', isEditorContextRequired)
        if (isCodebaseContextRequired || isEditorContextRequired) {
            contextMessages.push(...ChatQuestion.getEditorContext(editor))
        }

        // Add selected text as context when available
        if (selection?.selectedText) {
            contextMessages.push(...ChatQuestion.getEditorSelectionContext(editor.getActiveTextDocument()!, selection))
        }

        return contextMessages
    }

    public static getEditorContext(editor: Editor): ContextMessage[] {
        const currentWorkspace = editor.getActiveWorkspace()
        if (!currentWorkspace) {
            return []
        }

        const currentDocument = editor.getActiveTextDocument()
        if (!currentDocument?.visible) {
            return []
        }

        const filePath = currentWorkspace.relativeTo(currentDocument.uri)
        if (!filePath) {
            return []
        }

        const offset = new DocumentOffsets(currentDocument.content)

        const truncatedContent = truncateText(offset.jointRangeSlice(currentDocument.visible), MAX_CURRENT_FILE_TOKENS)
        return getContextMessageWithResponse(
            populateCurrentEditorContextTemplate(truncatedContent, filePath, currentDocument.repoName ?? undefined),
            {
                fileName: filePath,
                repoName: currentDocument.repoName ?? undefined,
                revision: currentDocument.revision ?? undefined,
            }
        )
    }

    public static getEditorSelectionContext(document: TextDocument, selection: SelectionText): ContextMessage[] {
        const filePath = uriToPath(document.uri)
        if (!filePath) {
            return []
        }

        const truncatedContent = truncateText(selection.selectedText, MAX_CURRENT_FILE_TOKENS)

        return getContextMessageWithResponse(
            populateCurrentEditorSelectedContextTemplate(truncatedContent, filePath, document.repoName ?? undefined),
            {
                fileName: filePath,
                ...selection,
            }
        )
    }
}
