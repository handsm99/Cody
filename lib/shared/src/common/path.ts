import type { URI } from 'vscode-uri'

import { isWindows as _isWindows } from './platform'

export interface PathFunctions {
    /**
     * All but the last element of path, or "." if that would be the empty path.
     */
    dirname: (path: string) => string

    /**
     * The last element of path, or "" if path is empty.
     * @param path the path to operate on
     * @param suffix optional suffix to remove
     */
    basename: (path: string, suffix?: string) => string

    /** The extension of path, including the last '.'. */
    extname: (path: string) => string

    /** Path separator. */
    separator: string
}

/** For file system paths on Windows ('\' separators and drive letters). */
export const windowsFilePaths: PathFunctions = pathFunctions(true)

/**
 * For POSIX and URI paths ('/' separators).
 */
export const posixAndURIPaths: PathFunctions & {
    /**
     * The relative path from {@link from} to {@link to}.
     *
     * Only implemented for POSIX and URI paths because there are currently no callers that need
     * this for Windows paths.
     */
    relative: (from: string, to: string) => string
} = { ...pathFunctions(false), relative: posixAndURIPathsRelative }

/**
 * Get the {@link PathFunctions} to use for the given URI's path.
 */
export function pathFunctionsForURI(uri: URI, isWindows = _isWindows()): PathFunctions {
    return uri.scheme === 'file' && isWindows ? windowsFilePaths : posixAndURIPaths
}

// I don't like reimplementing this here, but it's the best option because: (1) using Node's `path`
// module requires us to configure the bundler right or risk a subtle bug (and path-browserify
// doesn't have `path/win32` support, so we'd need multiple underlying packages); and (2)
// `vscode-uri` is hard to test because it has a global constant for the current platform set at
// init time.
function pathFunctions(isWindows: boolean): PathFunctions {
    const sep = isWindows ? '\\' : '/'
    const f: PathFunctions = {
        dirname(path: string): string {
            if (path === '') {
                return '.'
            }
            if (isWindows && isDriveLetter(path)) {
                return path
            }
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            if (path === '') {
                return sep
            }
            const i = path.lastIndexOf(sep)
            if (i === -1) {
                return '.'
            }
            if (i === 0) {
                return sep
            }
            path = path.slice(0, i)
            if (isWindows && isDriveLetter(path)) {
                return path + sep
            }
            return path
        },
        basename(path: string, suffix?: string): string {
            if (path.endsWith(sep)) {
                path = path.slice(0, -1)
            }
            if (isWindows && isDriveLetter(path)) {
                return ''
            }
            path = path.split(sep).at(-1) ?? ''
            if (suffix && path.endsWith(suffix)) {
                path = path.slice(0, -suffix.length)
            }
            return path
        },
        extname(path: string): string {
            const basename = f.basename(path)
            const i = basename.lastIndexOf('.')
            if (i === 0 || i === -1) {
                return ''
            }
            return basename.slice(i)
        },
        separator: sep,
    }
    return f
}

function isDriveLetter(path: string): boolean {
    return /^[A-Za-z]:$/.test(path)
}

/** The relative path from {@link from} to {@link to}. */
function posixAndURIPathsRelative(from: string, to: string): string {
    // Normalize slashes.
    from = from.replaceAll(/\/{2,}/g, '/')
    to = to.replaceAll(/\/{2,}/g, '/')

    // Trim trailing slashes.
    if (from !== '/' && from.endsWith('/')) {
        from = from.slice(0, -1)
    }
    if (to !== '/' && to.endsWith('/')) {
        to = to.slice(0, -1)
    }

    if (from === to) {
        return ''
    }

    // From a relative to an absolute path, the absolute path dominates.
    if (!from.startsWith('/') && to.startsWith('/')) {
        return to
    }

    const fromParts = from === '/' ? [''] : from.split('/')
    const toParts = to === '/' ? [''] : to.split('/')

    // Find the common root.
    let commonLength = 0
    while (
        commonLength < fromParts.length &&
        commonLength < toParts.length &&
        fromParts[commonLength] === toParts[commonLength]
    ) {
        commonLength++
    }

    // Calculate the number of directory traversals needed.
    const traversals = fromParts.length - commonLength

    // Build the relative path.
    const relativePath: string[] = []
    for (let i = 0; i < traversals; i++) {
        relativePath.push('..')
    }

    // Add the non-common path parts from the 'to' path.
    relativePath.push(...toParts.slice(commonLength))

    return relativePath.join('/')
}
