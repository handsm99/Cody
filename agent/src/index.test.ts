import assert from 'assert'
import { execSync, spawn } from 'child_process'
import path from 'path'

import { afterAll, describe, it } from 'vitest'

import { RecipeID } from '@sourcegraph/cody-shared/src/chat/recipes/recipe'

import { MessageHandler } from './jsonrpc'
import { ClientInfo } from './protocol'

export class TestClient extends MessageHandler {
    public async handshake(clientInfo: ClientInfo) {
        const info = await this.request('initialize', clientInfo)
        this.notify('initialized', null)
        return info
    }

    public listRecipes() {
        return this.request('recipes/list', null)
    }

    public async executeRecipe(id: RecipeID, humanChatInput: string) {
        return this.request('recipes/execute', {
            id,
            humanChatInput,
        })
    }

    public async shutdownAndExit() {
        await this.request('shutdown', null)
        this.notify('exit', null)
    }
}

describe.each([
    {
        name: 'FullConfig',
        clientInfo: {
            name: 'test-client',
            version: 'v1',
            workspaceRootUri: 'file:///path/to/foo',
            workspaceRootPath: '/path/to/foo',
            connectionConfiguration: {
                accessToken: process.env.SRC_ACCESS_TOKEN ?? 'invalid',
                serverEndpoint: process.env.SRC_ENDPOINT ?? 'invalid',
                customHeaders: {},
                autocompleteAdvancedProvider: 'anthropic',
                autocompleteAdvancedAccessToken: '',
                autocompleteAdvancedServerEndpoint: '',
                autocompleteAdvancedServerModel: null,
                autocompleteAdvancedEmbeddings: true,
                debug: false,
                verboseDebug: false,
            },
        },
    },
    {
        name: 'MinimalConfig',
        clientInfo: {
            name: 'test-client',
            version: 'v1',
            workspaceRootUri: 'file:///path/to/foo',
            workspaceRootPath: '/path/to/foo',
            connectionConfiguration: {
                accessToken: process.env.SRC_ACCESS_TOKEN ?? 'invalid',
                serverEndpoint: process.env.SRC_ENDPOINT ?? 'invalid',
                customHeaders: {},
            },
        },
    },
])('describe StandardAgent with $name', ({ clientInfo }) => {
    if (process.env.SRC_ACCESS_TOKEN === undefined || process.env.SRC_ENDPOINT === undefined) {
        it('no-op test because SRC_ACCESS_TOKEN is not set. To actually run the Cody Agent tests, set the environment variables SRC_ENDPOINT and SRC_ACCESS_TOKEN', () => {})
        return
    }
    const client = new TestClient()

    // Bundle the agent. When running `pnpm run test`, vitest doesn't re-run this step.
    execSync('pnpm run build')

    const agentProcess = spawn('node', ['--inspect', path.join(__dirname, '../dist/index.js'), '--inspect'], {
        stdio: 'pipe',
    })

    agentProcess.stdout.pipe(client.messageDecoder)
    client.messageEncoder.pipe(agentProcess.stdin)
    agentProcess.stderr.on('data', msg => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
        console.log(msg.toString())
    })

    it('initializes properly', async () => {
        const serverInfo = await client.handshake(clientInfo)
        assert.deepStrictEqual(serverInfo.name, 'cody-agent', 'Agent should be cody-agent')
        assert.deepStrictEqual(serverInfo.codyEnabled, true, 'Cody should be enabled')
    })

    it('lists recipes correctly', async () => {
        const recipes = await client.listRecipes()
        assert.equal(9, recipes.length)
    })

    it('returns non-empty autocomplete', async () => {
        const filePath = '/path/to/foo/file.js'
        const content = 'function sum(a, b) {\n    \n}'
        client.notify('textDocument/didOpen', { filePath, content })
        const completions = await client.request('autocomplete/execute', {
            filePath,
            position: { line: 1, character: 4 },
        })
        assert(completions.items.length > 0)
    })

    const streamingChatMessages = new Promise<void>(resolve => {
        client.registerNotification('chat/updateMessageInProgress', msg => {
            if (msg === null) {
                resolve()
            }
        })
    })

    it('allows us to execute recipes properly', async () => {
        await client.executeRecipe('chat-question', "What's 2+2?")
    })

    it('sends back transcript updates and makes sense', () => streamingChatMessages, 20_000)

    afterAll(async () => {
        await client.shutdownAndExit()
    })
})
