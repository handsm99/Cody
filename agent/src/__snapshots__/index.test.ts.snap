// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Agent > Chat > chat/submitMessage (long message) 1`] = `
"Absolutely! I'd be happy to generate a simple "Hello World" function in Java for you. Here's a straightforward implementation:

\`\`\`java:/src/HelloWorld.java
public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, World!");
    }
}
\`\`\`

This code defines a class called \`HelloWorld\` with a \`main\` method, which is the entry point of the Java program. The \`System.out.println()\` statement outputs the classic "Hello, World!" message to the console.

To compile and run this Java program, you can use the following commands in your terminal:

\`\`\`bash
javac /src/HelloWorld.java
\`\`\`

\`\`\`bash
java -cp /src HelloWorld
\`\`\`

These commands will compile the Java source file and then execute the compiled program, displaying the "Hello, World!" message on your screen. Enjoy your Java programming!"
`;

exports[`Agent > Chat > chat/submitMessage (with mock context) 1`] = `
"Certainly! Here's the Dog class that implements the Animal interface:

\`\`\`typescript:src/dog.ts
import { Animal } from './animal';

export class Dog implements Animal {
    name: string;
    isMammal: boolean = true;

    constructor(name: string) {
        this.name = name;
    }

    makeAnimalSound(): string {
        return "Woof!";
    }
}
\`\`\`

This implementation fulfills the Animal interface requirements and is ready to use in your workspace."
`;

exports[`Agent > Commands > commands/explain 1`] = `
"The code in @src/animal.ts:1-6 defines an interface called Animal. This interface serves as a blueprint for creating animal objects in a TypeScript program. Let's break down what this code does in simple terms:

1) The purpose of this code is to establish a common structure that all animal objects should follow. It defines the properties and methods that any animal object must have.

2) This interface doesn't take any inputs directly. Instead, it specifies what properties and methods should be present in any object that implements this interface.

3) The interface itself doesn't produce any outputs. However, it ensures that any object implementing this interface will have specific properties and methods that can be used to get information about the animal or perform certain actions.

4) The Animal interface achieves its purpose by declaring three members:
   - A 'name' property of type string, which would store the name of the animal.
   - A 'makeAnimalSound' method that returns a string, which would be used to represent the sound the animal makes.
   - An 'isMammal' property of type boolean, which would indicate whether the animal is a mammal or not.

5) While there's no complex logic or data transformation happening in this interface declaration, it sets up a structure that ensures consistency across different animal objects in the program. Any object that implements this interface must provide values for the 'name' and 'isMammal' properties, and implement the 'makeAnimalSound' method.

This interface is particularly useful for beginners as it introduces the concept of abstraction in programming. It allows you to define a common structure for different types of animals without specifying the exact details of how each animal will implement these properties and methods. For example, a 'Dog' class and a 'Cat' class could both implement this Animal interface, ensuring they both have a name, can make a sound, and specify whether they're mammals, while allowing for unique implementations of these features for each specific animal type."
`;

exports[`Agent > Commands > commands/smell 1`] = `
"After reviewing the code in @src/animal.ts:1-6, here are some potential improvements:

1. Add readonly modifier: Consider adding the \`readonly\` modifier to the \`name\` property if it's not intended to be changed after initialization. This can prevent accidental modifications and improve type safety.

\`\`\`typescript:src/animal.ts
export interface Animal {
    readonly name: string
    makeAnimalSound(): string
    isMammal: boolean
}
\`\`\`

2. Use method signature instead of property for \`isMammal\`: If the mammal status is determined dynamically, consider changing \`isMammal\` to a method signature. This allows for more flexibility in implementation.

\`\`\`typescript:src/animal.ts
export interface Animal {
    name: string
    makeAnimalSound(): string
    isMammal(): boolean
}
\`\`\`

3. Add JSDoc comments: Enhance documentation by adding JSDoc comments to describe the interface and its members. This improves code readability and provides better IDE support.

\`\`\`typescript:src/animal.ts
/**
 * Represents an animal with basic properties and behaviors.
 */
export interface Animal {
    /** The name of the animal */
    name: string
    /** Produces the sound associated with this animal */
    makeAnimalSound(): string
    /** Indicates whether the animal is a mammal */
    isMammal: boolean
}
\`\`\`

4. Consider using a type for the sound return: If there are specific types of animal sounds, you could define a union type for the return value of \`makeAnimalSound()\`. This can improve type safety and make the code more self-documenting.

\`\`\`typescript:src/animal.ts
type AnimalSound = 'bark' | 'meow' | 'roar' | 'chirp'

export interface Animal {
    name: string
    makeAnimalSound(): AnimalSound
    isMammal: boolean
}
\`\`\`

5. Add optional properties: If there are properties that might not apply to all animals, consider making them optional. This can make the interface more flexible and easier to implement for various animal types.

\`\`\`typescript:src/animal.ts
export interface Animal {
    name: string
    makeAnimalSound(): string
    isMammal: boolean
    habitat?: string
    lifespan?: number
}
\`\`\`

Overall, the current code follows sound design principles by using an interface to define a contract for animal objects. The suggestions provided aim to enhance type safety, flexibility, and documentation. While the existing code is functional, implementing these recommendations could lead to a more robust and maintainable codebase."
`;
